commit cf5648658c87b6e3b2fa4394b5921792827246dc
Author: Bryce Guinta <bryce.paul.guinta@gmail.com>
Date:   Sun Jan 7 14:28:42 2018 -0700

    Fix submodule import in six.moves

    This commit fixes import errors when modname
    started with, but was not equal, to six.moves
---
 astroid/brain/pysix_moves.py | 38 ++++++++++++++++++++++++++++++++----
 1 file changed, 34 insertions(+), 4 deletions(-)

---

The original patch have been amended to support the names of astroid 1.4.0
(because the patch was written for a more recent version of astroid) and
to also support six.moves.cPickle which was missing for python 2.


diff --git a/astroid/brain/brain_six.py b/astroid/brain/brain_six.py
index 9596a6c8..0935beb2 100644
--- a/astroid/brain/brain_six.py
+++ b/astroid/brain/brain_six.py
@@ -23,7 +23,7 @@ from textwrap import dedent
 
 from astroid import MANAGER, register_module_extender
 from astroid.builder import AstroidBuilder
-from astroid.exceptions import AstroidBuildingException, InferenceError
+from astroid.exceptions import AstroidBuildingException, InferenceError, NotFoundError
 from astroid import nodes
 
 
@@ -53,6 +53,7 @@ if sys.version_info[0] == 2:
     import CGIHTTPServer
     import SimpleHTTPServer
 
+    import cPickle
     from StringIO import StringIO
     from cStringIO import StringIO as cStringIO
     from UserDict import UserDict
@@ -199,7 +199,7 @@ else:
     import html.entities as html_entities
     import html.parser as html_parser
     import http.client as http_client
-    import http.server
+    import http.server as http_server
     BaseHTTPServer = CGIHTTPServer = SimpleHTTPServer = http.server
     import pickle as cPickle
     import queue
@@ -230,7 +230,8 @@ else:
     import tkinter.filedialog as tkinter_tkfiledialog
     import tkinter.font as tkinter_font
     import tkinter.messagebox as tkinter_messagebox
-    import urllib.request
+    import urllib
+    import urllib.request as urllib_request
     import urllib.robotparser as urllib_robotparser
     import urllib.parse as urllib_parse
     import urllib.error as urllib_error
@@ -253,10 +254,38 @@ def six_moves_transform():
 
 
 def _six_fail_hook(modname):
-    if modname != 'six.moves':
+    """Fix six.moves imports due to the dynamic nature of this
+    class.
+
+    Construct a psuedo-module which contains all the nessecary imports
+    for six
+
+    :param modname: Name of failed module
+    :type modname: str
+
+    :return: An astroid module
+    :rtype: nodes.Module
+    """
+
+    attribute_of = (modname != "six.moves" and
+                    modname.startswith("six.moves"))
+    if modname != 'six.moves' and not attribute_of:
         raise AstroidBuildingException
     module = AstroidBuilder(MANAGER).string_build(_IMPORTS)
     module.name = 'six.moves'
+    if attribute_of:
+        # Facilitate import of submodules in Moves
+        start_index = len(module.name)
+        attribute = modname[start_index:].lstrip(".").replace(".", "_")
+        try:
+            import_attr = module.getattr(attribute)[0]
+        except NotFoundError:
+            raise AstroidBuildingException
+        if isinstance(import_attr, nodes.Import):
+            submodule = MANAGER.ast_from_module_name(import_attr.names[0][0])
+            return submodule
+    # Let dummy submodule imports pass through
+    # This will cause an Uninferable result, which is okay
     return module
 
 def transform_six_add_metaclass(node):
